### PR TITLE
Integrate est prio

### PR DESCRIPTION
**Pull request by atatmaja**
## Briefly describe the feature or bug that this PR addresses.

- Integrates cleaned up version of estimate prioritization (written by @rahularoradfs) into our backend
- Adds code to test the new estimate prioritization script on real data
- Adds logic in the ETL to alert devteam if there are any columns that don't have pooling functions associated with them

## Please link the Airtable ticket associated with this PR.

https://airtable.com/tbli2lWQHAqBa6ZcI/viwWvxb27XGuziw4c/recHHgFj3YPNv3T4M

## Describe the steps you took to test the feature/bugfix introduced by this PR.

- Wrote the test mentioned above
- Made requests to the get records endpoint with several filter combinations

## Does any infrastructure work need to be done before this PR can be pushed to production?

No.

### CODE REVIEW

File: app/namespaces/data_provider/data_provider_controller.py [hunk 1]
```diff
@@ -92,7 +92,7 @@ def post(self):
         columns = data.get('columns')
         research_fields = data.get('research_fields')
         prioritize_estimates = data.get('prioritize_estimates', True)
-        prioritize_estimates_mode = data.get('prioritize_estimates_mode', None)
+        prioritize_estimates_mode = data.get('prioritize_estimates_mode', 'dashboard')
```
Comment (rahularoradfs) [inline app/namespaces/data_provider/data_provider_controller.py:95]: nice catch

File: app/utils/estimate_prioritization/__init__.py [hunk 2]
```diff
@@ -0,0 +1,3 @@
+from .pooling_functions import pooling_function_maps, get_columns_with_pooling_functions
+from .prioritization_criteria import prioritization_criteria_testadj, prioritization_criteria_testunadj
+from .estimate_prioritization import get_prioritized_estimates
```
Comment (rahularoradfs) [inline app/utils/estimate_prioritization/__init__.py]: no newline at end of file here - know this throws errors sometimes, so making commit to fix this

File: app/utils/estimate_prioritization/estimate_prioritization.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        # generate new values as necessary
+        pooled.at['numerator_value'] = int(pooled.at['serum_pos_prevalence'] * pooled.at['denominator_value'])
+        if 'estimate_name' in pooled.index:
+            pooled.at['estimate_name'] += '_pooled'
+        pooled.at['seroprev_95_ci_lower'], pooled.at['seroprev_95_ci_upper'] = proportion_confint(count = pooled.at['numerator_value'],
+                                                                                          nobs = pooled.at['denominator_value'],
+                                                                                          alpha = 0.05,
+                                                                                          method = 'jeffreys')
+
+        # TODO: once test adjustment is in, we'll need to add a parallel set of code to the above
```
Comment (rahularoradfs) [inline app/utils/estimate_prioritization/estimate_prioritization.py]: Making a commit to add the test adjustment pooling code here: ways to pool `[adj_prevalence, adj_sensitivity, adj_specificity, adj_prev_ci_lower, adj_prev_ci_upper, ind_eval_type]` Will wrap in try / except so we catch any keyerrors before merge or when the research table isn't included.

File: app/utils/estimate_prioritization/estimate_prioritization_test.py [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+
+    for var_name, expected in kwargs.items():
+        actual = result[var_name]
+        try:
+            assert expected == actual
+        except Exception as e:
+            # Need to catch this case because nan != nan
+            if not pd.isna(expected) and pd.isna(actual):
+                print(f"Failed test: {test_name}. Expected {expected}, actual {actual}")
+                raise e
```
Comment (rahularoradfs) [inline app/utils/estimate_prioritization/estimate_prioritization_test.py:23]: nice - very clean catch / reraise pattern

File: app/utils/estimate_prioritization/estimate_prioritization_test.py [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+            'age': 'All',
+            'sex': 'ALL',
+            'TEST_INFO_AVAIL': False
+        },
+    ]
+
+    # Adding these hardcoded values for now to make tests pass
+    for estimate in sample_estimates:
+        estimate['population_group'] = "General Population"
+        estimate['state'] = ["Texas"]
```
Comment (rahularoradfs) [inline app/utils/estimate_prioritization/estimate_prioritization_test.py:297]: yeehaw

File: app/utils/estimate_prioritization/estimate_prioritization_test.py [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+        estimate['city'] = ["Dallas"]
+        estimate['estimate_grade'] = 'National'
+
+    sample_df = pd.DataFrame(sample_estimates)
+    # Adding these hardcoded values for now to make tests pass
+    sample_df['sampling_start_date'] = datetime.datetime.now()
```
Comment (rahularoradfs) [inline app/utils/estimate_prioritization/estimate_prioritization_test.py:303]: might be simpler to add in the other hardcoded values at this point too - you could nix the for loop if you do it on a dataframe level

File: app/utils/estimate_prioritization/estimate_prioritization_test.py [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+                                test_name = 'SPECIMEN TYPE')
+
+    estimate_prioritization_test(sample_df,
+                                (sample_df['study_name'] == 'dynamic adjustment'),
+                                mode = 'analysis_dynamic',
+                                IDENTIFIER = 0,
+                                test_name = 'DYNAMIC ADJUSTMENT')
+
+# Test with real estimates queried via our endpoint
+def test_real_estimates():
```
Comment (rahularoradfs) [inline app/utils/estimate_prioritization/estimate_prioritization_test.py:359]: appreciate the download of real estimates to properly test this - thanks! Think this makes a lot of sense as a way to check the overall architecture. We're not testing individual functions here, but I think that's fine - to come up with a way to test those would be laborious and prone to breaking

**Review by rahularoradfs — APPROVED** [review]
minor comments. great code, thank you! you'll probably want to take a quick look at the commits I added to make sure they check out to you before merging!

File: app/utils/estimate_prioritization/estimate_prioritization.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
+import pandas as pd
+from statsmodels.stats.proportion import proportion_confint
+from app.utils.estimate_prioritization import prioritization_criteria_testunadj, \
+    prioritization_criteria_testadj, pooling_function_maps
+
+
+def get_pooled_estimate(estimates):
```
Comment (simonarocco) [inline app/utils/estimate_prioritization/estimate_prioritization.py:7]: add typing

File: app/database_etl/postgres_tables_handler/postgres_utils.py [hunk 9]
```diff
@@ -141,3 +142,31 @@ def check_filter_options(dashboard_source: pd.DataFrame) -> None:
...
+    missing_in_pooling = columns_in_db - columns_with_pooling_functions
+    missing_in_db = columns_with_pooling_functions - columns_in_db
+
+    if missing_in_pooling:
+        msg = f"The following columns are in the DB, but are missing pooling functions: {missing_in_pooling}"
+        send_slack_message(message=msg, channel='#dev-logging-etl')
+    if missing_in_db:
+        msg = f"The following columns have pooling functions, but are missing in the DB: {missing_in_db}"
+        send_slack_message(message=msg, channel='#dev-logging-etl')
+
```
Comment (simonarocco) [inline app/database_etl/postgres_tables_handler/postgres_utils.py:171]: remove blank line

File: app/database_etl/postgres_tables_handler/postgres_utils.py [hunk 10]
```diff
@@ -141,3 +142,31 @@ def check_filter_options(dashboard_source: pd.DataFrame) -> None:
         send_slack_message(message=body, channel='#dev-logging-etl')
 
     return
+
+
+# Send slack alert if there are any columns that are not accounted for
+# during the pooling process (in estimate prioritization)
+def validate_pooling_function_columns(tables_dict: Dict) -> None:
+    columns_with_pooling_functions = get_columns_with_pooling_functions()
+    columns_in_db = set.union(*[set(table.columns) for table in tables_dict.values()])
```
Comment (simonarocco) [inline app/database_etl/postgres_tables_handler/postgres_utils.py:151]: add link to documentation for this plz

File: app/utils/estimate_prioritization/estimate_prioritization.py [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+        # use the estimate with the max denominator as the 'base' for the pooled estimate
+        pooled = estimates.loc[estimates['denominator_value'].idxmax(axis = 1)] 
+
+        # pool all values that can be pooled
+        for (summary_type, cols_to_summarize, summary_function) in pooling_function_maps:
+            if summary_function is not None:
+                valid_cols_to_summarize = set(cols_to_summarize) & set(estimates.columns)
+
+                for col in valid_cols_to_summarize:
+                    pooled.at[col] = summary_function(estimates, col)
```
Comment (simonarocco) [inline app/utils/estimate_prioritization/estimate_prioritization.py:21]: why .at and not .loc?

File: app/utils/estimate_prioritization/estimate_prioritization.py [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+
+def get_pooled_estimate(estimates):
+
+    if estimates.shape[0] == 1:
+        pooled = estimates.iloc[0]
+    else:
+        # use the estimate with the max denominator as the 'base' for the pooled estimate
+        pooled = estimates.loc[estimates['denominator_value'].idxmax(axis = 1)] 
+
+        # pool all values that can be pooled
```
Comment (simonarocco) [inline app/utils/estimate_prioritization/estimate_prioritization.py:15]: can we add better comments please from line 15 - 40? lacking comments and comments that exist are weak :/

File: app/utils/estimate_prioritization/estimate_prioritization.py [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+        pooled.at['seroprev_95_ci_lower'], pooled.at['seroprev_95_ci_upper'] = proportion_confint(count = pooled.at['numerator_value'],
+                                                                                          nobs = pooled.at['denominator_value'],
+                                                                                          alpha = 0.05,
+                                                                                          method = 'jeffreys')
+        
+        try:
+            pooled.at['adj_prev_ci_lower'], pooled.at['adj_prev_ci_upper'] = proportion_confint(count = int(pooled.at['adj_prevalence'] * pooled.at['denominator_value']),
+                                                                                                nobs = pooled.at['denominator_value'],
+                                                                                                alpha = 0.05,
+                                                                                                method = 'jeffreys')
```
Comment (simonarocco) [inline app/utils/estimate_prioritization/estimate_prioritization.py:36]: remove blank line

File: app/utils/estimate_prioritization/estimate_prioritization.py [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+        except KeyError as e: 
+            pass
+
```
Comment (simonarocco) [inline app/utils/estimate_prioritization/estimate_prioritization.py:39]: remove blank line

File: app/utils/estimate_prioritization/estimate_prioritization.py [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+    return pooled
+
+# pass in a filtered set of estimates - or estimates and filters
+# and get out a subset, which have an estimate prioritized from among them 
+def get_prioritized_estimates(estimates,
```
Comment (simonarocco) [inline app/utils/estimate_prioritization/estimate_prioritization.py:44]: add typing

File: app/utils/estimate_prioritization/estimate_prioritization.py [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+    return pooled
+
+# pass in a filtered set of estimates - or estimates and filters
+# and get out a subset, which have an estimate prioritized from among them 
+def get_prioritized_estimates(estimates, 
+                              filters = None, 
+                              mode = 'dashboard',
+                              pool = True): 
+    
+    if estimates.shape[0] == 0:
```
Comment (simonarocco) [inline app/utils/estimate_prioritization/estimate_prioritization.py:49]: if estimates.empty

File: app/utils/estimate_prioritization/prioritization_criteria.py [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+        lambda estimate: estimate['test_type'] == 'ELISA'
+    ],
+    'specimen': [
+        lambda estimate: estimate['specimen_type'] != 'Dried Blood'
+    ]
+}
+
+prioritization_criteria_testadj = {name: criterion for name, criterion
+                                  in prioritization_criteria_full.items()
+                                  if name.find('testunadj') == -1}
```
Comment (simonarocco) [inline app/utils/estimate_prioritization/prioritization_criteria.py:66]: if not 'testunadj' in name?

File: app/utils/estimate_prioritization/estimate_prioritization.py [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+            # if a prioritization criterion yields multiple estimates, continue applying criteria 
+
+        # select a criterion
+        for criterion_name, criterion in prioritization_criteria.items():
+            # go through the levels of that criterion, from highest to lowest
+            for level in criterion:
+                study_estimates_at_level = study_estimates[study_estimates.apply(level, axis = 1)]
+
+                # if any estimates meet the top level, pass those on; if not, continue
+                if study_estimates_at_level.shape[0] > 0:
```
Comment (simonarocco) [inline app/utils/estimate_prioritization/estimate_prioritization.py:98]: if not study_estimates_at_level.empty

File: app/utils/estimate_prioritization/estimate_prioritization_test.py [hunk 19]
```diff
@@ -0,0 +1,396 @@
+from app.utils.estimate_prioritization import get_prioritized_estimates
+import datetime
+import pandas as pd
+import json
+
+# test estimate prioritization code on that minimum set
+def estimate_prioritization_test(estimates,
```
Comment (simonarocco) [inline app/utils/estimate_prioritization/estimate_prioritization_test.py:7]: add typing

File: app/utils/estimate_prioritization/estimate_prioritization_test.py [hunk 20]
```diff
@@ -0,0 +1,396 @@
...
+            'test_type': 'ELISA',
+            'specimen_type': 'Not Dried Blood',
+            'age': 'All',
+            'sex': 'ALL',
+            'TEST_INFO_AVAIL': False
+        },
+    ]
+
+    # Adding these hardcoded values for now to make tests pass
+    for estimate in sample_estimates:
```
Comment (simonarocco) [inline app/utils/estimate_prioritization/estimate_prioritization_test.py:295]: why not just add these extra values to the sample estimates dict?

File: app/utils/estimate_prioritization/estimate_prioritization_test.py (hunk 6 above)
Comment (simonarocco) [inline app/utils/estimate_prioritization/estimate_prioritization_test.py:303]: same comment as above

File: app/utils/estimate_prioritization/pooling_functions.py [hunk 21]
```diff
@@ -0,0 +1,158 @@
...
+
+# we will need to 'pool' estimates, collapsing values from multiple estimates to one
+# here are functions that do that for each variable
+PoolingFnMap = namedtuple(typename = 'PoolingFnMap',
+                          field_names = ['summary_type',
+                                         'column_names',
+                                         'summary_function'])
+
+# helper function to use to generate lambdas
+def get_unique_value(series, default = pd.NA):
```
Comment (simonarocco) [inline app/utils/estimate_prioritization/pooling_functions.py:12]: add typing

**Review by simonarocco — COMMENTED** [review]
good work! suggestions to add typing for functions, more comments, and syntax alternatives

File: app/utils/estimate_prioritization/estimate_prioritization.py (hunk 11 above)
Comment (atatmaja) [reply app/utils/estimate_prioritization/estimate_prioritization.py:21]: pooled is a series and not a DF. @rahularoradfs can you confirm

File: app/utils/estimate_prioritization/estimate_prioritization.py (hunk 12 above)
Comment (atatmaja) [reply app/utils/estimate_prioritization/estimate_prioritization.py:15]: @rahularoradfs care to take a crack

File: app/utils/estimate_prioritization/estimate_prioritization.py (hunk 11 above)
Comment (rahularoradfs) [reply app/utils/estimate_prioritization/estimate_prioritization.py:21]: confirmed

File: app/utils/estimate_prioritization/estimate_prioritization.py (hunk 12 above)
Comment (rahularoradfs) [reply app/utils/estimate_prioritization/estimate_prioritization.py:15]: taking a crack now, give me 10

Comment (rahularoradfs) [reply app/utils/estimate_prioritization/estimate_prioritization.py:15]: "pool all values that can be pooled"

File: app/utils/estimate_prioritization/prioritization_criteria.py (hunk 17 above)
Comment (rahularoradfs) [reply app/utils/estimate_prioritization/prioritization_criteria.py:66]: name is a string - does in work in that case?

**Comment by rahularoradfs** [comment]
added comments to pooling function in `estimate_prioritization.py` per simona's req and austin's delegation!

File: app/utils/estimate_prioritization/estimate_prioritization.py (hunk 12 above)
Comment (rahularoradfs) [reply app/utils/estimate_prioritization/estimate_prioritization.py]: done

File: app/utils/get_filtered_records.py [hunk 22]
```diff
@@ -190,7 +190,15 @@ def date_filter(record, start_date=None, end_date=None, use_sampling_date=True):
     if prioritize_estimates:
         result_df = pd.DataFrame(result)
         prioritized_records = get_prioritized_estimates(result_df, mode=prioritize_estimates_mode)
+        # If records exist, clean dataframe
         if not prioritized_records.empty:
+            # Convert from True/None to True/False
+            # TODO: maybe this is something we enforce in the ETL moving fwd
+            bool_fields = ['academic_primary_estimate', 'dashboard_primary_estimate', 'test_adj', 'pop_adj', 'geo_exact_match']
```
Comment (rahularoradfs) [inline app/utils/get_filtered_records.py:197]: Is it worth it to generate bool fields automatically, so this doesn't break later if we add another bool field?

Could iterate over the columns of prioritized_records to see which have boolean dtypes, and apply this transformation to them

Comment (atatmaja) [reply app/utils/get_filtered_records.py:197]: yeah i think this is better

**Pull request merged**